### PR TITLE
Tweak completion filtering behavior. #2

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems?id=362890

This is a more extensive change thatn https://github.com/dotnet/roslyn/pull/16892.  Here, we also remove the 'recheckCaretPosition' check since that can be inferred from a FilterReason of 'CaretPositionChanged'.

### CODE REVIEW

File: src/Features/Core/Portable/Completion/CompletionTriggerKind.cs [hunk 1]
```diff
@@ -25,6 +28,18 @@ public enum CompletionTriggerKind
...
+        /// Completion was trigger by a direct invocation of the completion feature 
+        /// (ctrl-j in Visual Studio).
+        /// </summary>
+        Invoke,
+
+        /// <summary>
+        /// Completion was triggered with a request to commit if a unique item would be selected 
+        /// (ctrl-space in Visual Studio).
+        /// </summary>
+        InvokeAndCommitIfUnique
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionTriggerKind.cs]: Feedback welcome on the names 'Invoke' and 'InvokeAndCommitIfUnique'.

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide This simplifies some of our code around filtering completion lists, moving as much as possible to one location, and providing a very simple API with very few knobs for consumers to get wrong.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs [hunk 2]
```diff
@@ -194,36 +185,30 @@ internal override void OnModelUpdated(Model modelOpt)
...
-            }
-            else
-            {
-                sessionOpt.IdentifyBestMatchAndFilterToAllItems(
-                    filterReason,
-                    recheckCaretPosition: false,
-                    dismissIfEmptyAllowed: dismissIfEmptyAllowed);
-            }
-        }
+        //private void FilterToSomeOrAllItems(bool filterItems, bool dismissIfEmptyAllowed, CompletionFilterReason filterReason)
```
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: Remove?

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: I think you're looking at an old review :)

File: src/Features/Core/Portable/Completion/CompletionTriggerKind.cs (hunk 1 above)
Comment (rchande) [inline src/Features/Core/Portable/Completion/CompletionTriggerKind.cs]: The VS commands are called `list members` and `complete word`. I like `complete word` but not `list members`.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_OnTextViewBufferPostChanged.cs [hunk 3]
```diff
@@ -29,12 +29,14 @@ internal override void OnTextViewBufferPostChanged(object sender, EventArgs e)
...
-                    recheckCaretPosition: model == null,
-                    dismissIfEmptyAllowed: true,
-                    filterState: null);
+                // Something changed in the buffer without us hearing about the change first.
+                // This can happen in complex projection scenarios (with buffers being mapped/unmapped).
+                // In this case, queue up a CaretPositionChanged filter first.  That way if the caret
+                // moved out of bounds of the items, then we'll dismiss.  Also queue up an insertion.
+                // that way we go and filter things properly to ensure that the list contains the 
+                // appropriate items.
+                sessionOpt.FilterModel(CompletionFilterReason.CaretPositionChanged, filterState: null);
```
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_OnTextViewBufferPostChanged.cs]: nice, lots of cryptic flags removed.

**Pull request merged**